sklearn/utils/multiclass.py

```diff
@@ -387,7 +387,17 @@ def type_of_target(y, input_name=""):
 
     # Check multiclass
     first_row = y[0] if not issparse(y) else y.getrow(0).data
-    if xp.unique_values(y).shape[0] > 2 or (y.ndim == 2 and len(first_row) > 1):
+    classes = xp.unique_values(y)
+    if classes.shape[0] > round(0.5 * y.shape[0]) and classes.shape[0] > 1:
+        warnings.formatwarning(
+            message=r"The number of unique classes is greater than 50% of the samples.",
+            category=UserWarning,
+            filename="sklearn/utils/tests/test_multiclass.py",
+            lineno=-1,
+            line="type_of_target(y)",
+        )
```

Comment:
Could you explain a bit why `warnings.formatwarning` is needed instead of `warnings.warn`? I was expecting to see something like the following:
```suggestion
        warnings.warn(
            "The number of unique classes is greater than 50% of the samples.",
            UserWarning,
            stacklevel=2,
        )
```